### PR TITLE
feat: Allows user to pass in a pointer to their struct so that they don't have to deal with raw []byte objects

### PR DESCRIPTION
**Pull request by John-Memphis**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -112,23 +153,28 @@ type Event struct {
 	Field2 string `json:"field2"`
 }
 
-func eventHandlerFunc(msgPayload[]byte, msgHeaders map[string]string, inputs map[string]string) ([]byte, map[string]string, error){
+func eventHandlerFunc(msgPayload any, msgHeaders map[string]string, inputs map[string]string) ([]byte, map[string]string, error){
+    // Type assert to user type. By default this is a *[]byte
+    as_bytes, ok := msgPayload.(*[]byte)
```
Comment (shay23b) [inline README.md]: change *[]byte to []byte

File: memphis/memphis.go [hunk 2]
```diff
@@ -28,17 +31,77 @@ type MemphisOutput struct {
...
+	// Unmarshal JSON data into the struct
+	err := json.Unmarshal(data, userStruct)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checkValidStruct(userStruct any) error {
```
Comment (shay23b) [inline memphis/memphis.go]: can remove this validation - it's ok to fail on the Unmarshal function if the user didn't pass the right format

File: memphis/memphis.go [hunk 3]
```diff
@@ -28,17 +31,77 @@ type MemphisOutput struct {
...
+		for _, option := range options {
+			if option != nil {
+				if err := option(&params); err != nil {
+					return nil, err
+				}
+			}
+		}
+
+		if params.UserObject != nil {
+			err := checkValidStruct(params.UserObject)
```
Comment (shay23b) [inline memphis/memphis.go]: unnecessary

File: memphis/memphis.go [hunk 4]
```diff
@@ -28,17 +31,77 @@ type MemphisOutput struct {
...
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		if len(options) > 1 {
```
Comment (shay23b) [inline memphis/memphis.go]: unnecessary

File: memphis/memphis.go [hunk 5]
```diff
@@ -28,17 +31,77 @@ type MemphisOutput struct {
...
 // if all returned values are nil the message will be filtered out from the station.
-func CreateFunction(eventHandler EventHandlerFunction) {
+func CreateFunction(eventHandler HandlerType, options ...HandlerOption) {
 	LambdaHandler := func(ctx context.Context, event *MemphisEvent) (*MemphisOutput, error) {
+		params := HandlerOptions{
+			Handler:           eventHandler,
+			UserObject:        nil,
+		}
+
+		for _, option := range options {
```
Comment (shay23b) [inline memphis/memphis.go:84]: move to before LambdaHandler and pass it inside

**Review by shay23b — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (John-Memphis) [reply README.md]: https://github.com/memphisdev/memphis-functions.go/pull/6/commits/19bc8ecb18e828936b92ad5c0db770af4fb0bf8d

Done

File: memphis/memphis.go (hunk 5 above)
Comment (John-Memphis) [reply memphis/memphis.go:84]: The lambda handler is given to the lambda.Start, and must follow their  signature:
https://docs.aws.amazon.com/lambda/latest/dg/golang-handler.html

> func HandleRequest(ctx context.Context, event *MyEvent) (*string, error): This is the signature of your Lambda handler. It's the entry point for your Lambda function and contains the logic that is executed when your function is invoked. In addition, the parameters included denote the following:

In order for Lambda to work, this must maintain the same. I cannot pass these in the LambdaHandler, they muse be used like this.

File: memphis/memphis.go (hunk 4 above)
Comment (John-Memphis) [reply memphis/memphis.go]: https://github.com/memphisdev/memphis-functions.go/pull/6/commits/8df5a62c5dd353cfc9ccbcadeb70eecf1e288de9

File: memphis/memphis.go (hunk 3 above)
Comment (John-Memphis) [reply memphis/memphis.go]: https://github.com/memphisdev/memphis-functions.go/pull/6/commits/8df5a62c5dd353cfc9ccbcadeb70eecf1e288de9

File: memphis/memphis.go (hunk 2 above)
Comment (John-Memphis) [reply memphis/memphis.go]: https://github.com/memphisdev/memphis-functions.go/pull/6/commits/8df5a62c5dd353cfc9ccbcadeb70eecf1e288de9

File: memphis/memphis.go [hunk 6]
```diff
@@ -28,17 +29,66 @@ type MemphisOutput struct {
...
+// message headers as map[string]string and inputs as map[string]string and should return the modified payload and headers.
+// The modified payload type will either be the user type, or []byte depending on user requirements.
 // error should be returned if the message should be considered failed and go into the dead-letter station.
 // if all returned values are nil the message will be filtered out from the station.
-func CreateFunction(eventHandler EventHandlerFunction) {
+func CreateFunction(eventHandler HandlerType, options ...PayloadOption) {
 	LambdaHandler := func(ctx context.Context, event *MemphisEvent) (*MemphisOutput, error) {
+		params := PayloadOptions{
+			Handler:    eventHandler,
+			UserObject: nil,
```
Comment (shay23b) [inline memphis/memphis.go]: Let's name it UserStruct instead, in Go using 'object' wouldn't be quite right

**Review by shay23b — CHANGES_REQUESTED** [review]

File: memphis/memphis.go (hunk 6 above)
Comment (John-Memphis) [reply memphis/memphis.go]: I renamed it

**Review by shay23b — APPROVED** [review]

**Pull request merged**